src/readbin/compare.ml

```diff
@@ -0,0 +1,106 @@
+open Core_kernel.Std
+open Cmdliner
+open Bap.Std
+open Bap_plugins.Std
+
+exception Not_found
+
+let print_source = function
+  | `bw -> print_endline "bw"
+  | _ -> print_endline "else"
+
+let source = [
+  "BW", `bw;
+  "SymTbl", `symtbl;
+  "User", `user;
+  "Ida", `ida;
+]
+
+let tool : _ Term.t =
+  let doc = "The tool that we are going to evaluate" in
+  Arg.(value & pos 1 (enum source) `bw & info [] ~docv:"tool" ~doc)
+
+let gt : _ Term.t =
+  let doc = "The ground truth source. One can direct to a user file, the symbol
+  table" in
+  Arg.(value & pos 2 (enum source) `symtbl & info [] ~docv:"gt" ~doc)
+
+let symsfile : string option Term.t =
+  let doc = "The symbol table of binaries. This requires the binaries Use this file as symbols source." in
+  Arg.(value & opt (some non_dir_file) None & info ["syms"; "s"]
+         ~docv:"syms" ~doc)
+
+let print_metrics : _ list Term.t =
+  let opts = [
+    "precision", `with_prec;
+    "recall", `with_recl;
+    "F_measure", `with_F;
+    "TP", `with_TP;
+    "FP", `with_FP;
+    "FN", `with_FN;
+  ] in
+  let doc = "doc" in
+  Arg.(value & opt_all (enum opts) (List.map ~f:snd opts) &
+       info ["with-metrics"; "c"] ~doc)
+
+let bin : string Term.t =
+  let doc = "binary" in
+  Arg.(required & pos 0 (some non_dir_file) None & info [] ~docv:"binary" ~doc)
+
+let func_start bin symsfile : _ -> Addr.Hash_set.t * string = function
+  | `bw -> Func_start.byteweight bin, "BW"
+  | `user -> (match symsfile with
+      | None -> raise Not_found
+      | Some f -> Func_start.usersource f, "User")
+  | `symtbl -> Func_start.symbols bin, "Symbol"
+  | `ida -> Func_start.ida bin, "IDA"
+  | _ -> Addr.Hash_set.create (), "None"
+
+let compare bin print_metrics tool gt symsfile : unit = try
+    let fs_tool, tool_name = func_start bin symsfile tool in
+    let fs_gt, _ = func_start bin symsfile gt in
+    let fp =
+      let set = Hash_set.diff fs_tool fs_gt in
+      Hash_set.length set in
+    let fn =
+      let set = Hash_set.diff fs_gt fs_tool in
+      Hash_set.length set in
+    let tp = Hash_set.length fs_gt - fn in
```

Comment:
consider factoring out the calculation logic out of the `compare` function. I would even put it into a separate module with careful documentation, since this is the code the people will try to read, in order to figure out, what your statistics mean.